### ISSUE TITLE
add passwordcredentials client like clientcredentials

### ISSUE THREAD

**Issue by josephholsten** [golang/oauth2#186]
https://go-review.googlesource.com/c/23834/

In order to use PasswordCredentialsToken, a TokenSource needs to be provided. It's not complicated to have one of these readily available, in the style of `golang.org/x/oauth2/clientcredentials`.

It would also be good to have #146, which would fix #110.

**Comment by josephholsten** [comment]
CL in http://golang.org/cl/23611/

**Comment by josephholsten** [comment]
needs tests and doc

**Comment by josephholsten** [comment]
Tests done, doc needs review

**Comment by josephholsten** [comment]
doc should be ready

**Comment by josephholsten** [comment]
reverted by https://go-review.googlesource.com/#/c/23841/, apparently this requires more discussion

**Comment by josephholsten** [comment]
@rakyll I added another comment to the CL, you made a great point. That's how I ended up submitting this. This "reimplements" https://godoc.org/golang.org/x/oauth2#Config.PasswordCredentialsToken using the same implementation pattern from https://godoc.org/golang.org/x/oauth2/clientcredentials, because using PasswordCredentialsToken as is requires implementing a wrapper TokenSource.

I'm planning to make a CL reimplementing this using Config.PasswordCredentialsToken. That was how I started in my local fork and didn't like it as much. I don't know if/how to re-open this CL, but I still think it's the most reasonable.

**Comment by adg** [comment]
This CL is still pending, right? https://go-review.googlesource.com/c/23834/

**Comment by rakyll** [comment]
Even though we promised not to break any APIs, I am suggesting to remove the Config.PasswordCredentialsToken if we merge the passwordcredentials package. I do not think the current API is widely used and keeping the library orthogonal is more valuable than breaking a few users.

**Comment by josephholsten** [comment]
Yes, I am still using this approach locally. I'm still interested in this CL.

**Comment by josephholsten** [comment]
@adg is there anything I need to do to move this forward?

**Comment by adg** [comment]
@rakyll what do you think?

**Comment by rakyll** [comment]
I am in favor of adding a deprecation notice to https://godoc.org/golang.org/x/oauth2#Config.PasswordCredentialsToken and go with @josephholsten's package.

**Comment by josephholsten** [comment]
@rakyll @adg anything you'd like me to do?

**Comment by adg** [comment]
What @rakyll suggests sounds good to me. Whoever wants to do that can do it.

**Comment by rakyll** [comment]
@josephholsten, I left comments on your change at https://go-review.googlesource.com/c/23834/ a while ago. Can you take a look at those?

**Comment by josephholsten** [comment]
@rakyll sorry, I hadn't pushed my changes when I made that comment. Care to give it another look?

**Comment by smitt04** [comment]
I would like to use this package, is this every going to move forward?

**Comment by joefitzgerald** [comment]
For anyone who is interested, a couple of years ago I extracted this PR into a package that you can use to provide the required functionality, pending the merge of this PR: https://github.com/joefitzgerald/passwordcredentials

**Comment by seankhliao** [comment]
We're not going to support this.
This is disallowed by RFC 9700 https://datatracker.ietf.org/doc/html/rfc9700#name-resource-owner-password-cre and not part of OAuth 2.1

**Comment by josephholsten** [comment]
@seankhliao if you intend to comply with the recommendations in RFC9700, we’ll need a change to https://godoc.org/golang.org/x/oauth2#Config.PasswordCredentialsToken, either to remove or mark deprecated. Merely choosing not to maintain that code doesn’t actually improve the security posture of the code.

That being said, the bikeshedding on this contribution has been hilariously overwrought. Do with this code as you will, I’ve long moved on with my life.
